src/translator/byte_array_util.cpp

```diff
@@ -104,5 +104,15 @@ AlignedMemory getShortlistMemoryFromConfig(marian::Ptr<marian::Options> options)
   return loadFileToMemory(shortlist[0], 64);
 }
 
+void getVocabsMemoryFromConfig(marian::Ptr<marian::Options> options,
+                               std::vector<AlignedMemory>& vocabMemorySet){
+  auto vfiles = options->get<std::vector<std::string>>("vocabs");
+  ABORT_IF(vfiles.size() < 2, "Insufficient number of vocabularies.");
+  vocabMemorySet.resize(vfiles.size());
+  for (size_t i = 0; i < vfiles.size(); ++i) {
+    vocabMemorySet[i] = loadFileToMemory(vfiles[i], 64);
+  }
```

Comment:
@kpu This is not a set, @qianqianzhu is double loading it even if the files are same.

Reply:
Here is my current struggle. I tried to avoid duplicate loading but failed. As here we need to provide the number of vocabularies. If you have any ideas, please tell me.

Reply:
This is tricky territory. 

I think we should take from the callee (Mozilla) a `vector<AlignedMemory> vocabMemories` of unique vocabulary memories and a `vector<size_t> vocabIndices` indicating the index where we load (each vocabs) to not double copy the memory (also not double load).  

So if we're reusing source and target vocabularies as `esen`, we'll have `vocabMemories.size() = 1`, and `vocabIndices.size() = 2`.

With this we can build a replicate the current hashmap behaviour with filenames on `vocabIndices` avoiding duplication between loading from files/memory as well I think.

/cc @abhi-agg

Reply:
@jerinphilip many thanks for your suggestion. (1) Introducing `vector<size_t> vocabIndices` is indeed a good way to avoid double loading.

Following your thoughts, if we will have a pair of vocabularies (source and target), we can just deal with `vocabMemories.size() = 1` and `vocabMemories.size() = 2` separately. (2) When `vocabMemories.size() = 1`, we initialise source and target vocabs from the same buffer. When `size = 2`, we initialise source and target vocabs from different buffers.

Or we can (3) use `std::vector<marian::Ptr<AlignedMemory>> vocabMemories` to avoid double loading. I hope introducing `shared_ptr` will not affect the WASM JS binding (@abhi-agg). 

@kpu @jerinphilip @abhi-agg Let me know which way works best.

Reply:
@qianqianzhu: 2 questions:

1. Does marian avoid loading the vocabularies twice into memory when both the vocabulary files passed to its command line parameter `--vocabs` are the same?

2. If yes, how does it decide to do so? By checking whether both file names are same?

From the API consumer perspective, I am probably missing the point of having `vocabIndices`. Providing single entry in `vocabMemories` but 2 in `vocabIndices` mean the consumer of the API should be aware that the 2 vocab files are same and hence create only 1 AlignedMemory object. If the API consumer is expected to know this then probably we could get away with `vocabIndices` completely. Would following work:

1. When both vocab files are exactly the same (e.g. `esen` case), the API consumer will create only 1 `AlignedMemory` object, create a vector with this object and call the `Service` constructor.

2. When both vocab files are different, the consumer will create 2 `AlignedMemory` objects, create a vector with these 2 objects and call the `Service` constructor.

Let me know if I missed something here 👍
cc @jerinphilip @kpu

Reply:
Sorry I missed this comment. 

@abhi-agg For marian, it does not care whether the vocabularies are the same or not. It just loads the files even they are the same. You can see here: https://github.com/browsermt/marian-dev/blob/master/src/translator/translator.h#L231-L238

For bergamot-translator, we did optimise the loading part: we avoid double loading by checking the filenames. https://github.com/browsermt/bergamot-translator/blob/main/src/translator/service.cpp#L19-L22

Reply:
Need some help here to make a decision. @kpu @jerinphilip

Reply:
@qianqianzhu still an issue, or did you choose a solution?

Reply:
I will go for `std::vector<marian::Ptr<AlignedMemory>>` if it is ok @XapaJIaMnu.

Reply:
@qianqianzhu Would passing just `std::vector<AlignedMemory>` not work?

The API consumer will pass `std::vector<AlignedMemory>` containing unique AlignedMemory objects.
Listing down some scenarios:

Scenario 1:
Source vocabs = "vocab1", "vocab2", "vocab3"
Target vocabs = "vocab4"

You will have 4 unique entries in `std::vector<AlignedMemory>`

Scenario 2:
Source vocabs = "vocab1"
Target vocabs = "vocab1"

You will have 1 unique entry in `std::vector<AlignedMemory>`

Did I miss any other scenario? If not then this solution should work?

Reply:
@abhi-agg you are right. Thought `shared_ptr` would be easier to manage the objects.

Reply:
Wait. @abhi-agg How about the following scenario:
Source vocabs = "vocab1", "vocab2", "vocab1"
Target vocabs = "vocab1"

Reply:
@qianqianzhu I don't have any issue with using `shared_ptr` in the API but this discussion opened up a topic for clarification that probably you, @kpu, @XapaJIaMnu or @jerinphilip can answer :)

1. Can marian work with multiple source vocabularies?
    1. Reason of confusion: The [code](https://github.com/browsermt/marian-dev/blob/master/src/translator/translator.h#L231-L238) + [data members](https://github.com/browsermt/marian-dev/blob/master/src/translator/translator.h#L213-L214) lead me to believe that yes it is possible. However, I couldn't find any example in [marian documentation](https://marian-nmt.github.io/docs/) that uses multiple source vocabularies. The [--vocabs option](https://github.com/browsermt/marian-dev/blob/master/src/common/config_parser.cpp#L643-L644) is indeed a vector<string> but it is not mentioned anywhere whether the user can pass more than 2 entries for it.
 
2. I think the answer to this one is "No" but for the sake of completion asking here, can marian work with multiple target vocabularies?

If marian is not supposed to work with multiple source vocabularies then we don't need to use `shared_ptr` in the API because it wouldn't be needed. If both source and target vocab is same, `vector<AlignedMemory>` will contain single entry otherwise 2 entries :) However, I am still fine with `shared_ptr` 👍

Reply:
I think we should use `shared_ptr` (and not marian::Ptr which is the same thing but exposes a Marian internal).  

Marian can use different vocabularies for source and target.  More generally, Marian supports multiple inputs each of which can have its own vocabulary.  This has been used for example to do http://statmt.org/wmt21/triangular-mt-task.html .

Reply:
>  Marian can use different vocabularies for source and target. More generally, Marian supports multiple inputs each of which can have its own vocabulary.

This clears few things. Thanks.

For the case when the user specifies only 1 entry for `--input` option, can the user specify multiple source and target vocabularies?

Reply:
> For the case when the user specifies only 1 entry for --input option, can the user specify multiple source and target vocabularies?

I don't know. But my guessing is no as one input corresponds to one vocabulary. Multiple source vocabularies for one input source seems not necessary. @XapaJIaMnu knows better than me.

Reply:
Every system you will encounter @abhi-agg has exactly two vocabulary slots: one for source and one for target.  These vocabularies can be the same.  

If you provide a single vocabulary slot, shame on you that's an error.

Reply:
> Every system you will encounter @abhi-agg has exactly two vocabulary slots: one for source and one for target. These vocabularies can be the same.
> 
> It essentially never makes sense to have multiple vocabularies for the same input and anyway Marian doesn't support it.

This explanation clears everything for me. Thanks again 👍